### PR TITLE
RUN-3390 (phase 1) separate plugins from preload scripts

### PR DESCRIPTION
**Pull request by whyn07m3**
This is **Phase 1** of _plugin-preload separation/refactor/cleanup/fix_.

All plugin/preload tests will be written when all phases are completed.

This 1st phase:
• separates plugin/preload logic in api-decorator and window handler files
• ensures plugins are eval'ed 1st
• fixes preload script updates to final states

✅ Test results:
• [Windows 7](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59e7bae7af26a25be2ffc8f9)
• [Windows 10](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59e7bb7baf26a25be2ffc8fa)

### CODE REVIEW

File: src/browser/api/window.js [hunk 1]
```diff
@@ -794,6 +794,8 @@ Window.create = function(id, opts) {
         _window: browserWindow
     };
 
+    winObj.pluginState = []; // TODO
```
[the remark this answers is not in the document]
Comment (whyn07m3) [reply src/browser/api/window.js]: this is taken care of in the last phase...

File: src/browser/api/window.js [hunk 2]
```diff
@@ -1137,26 +1140,39 @@ Window.getParentApplication = function() {
...
+ */
+Window.setWindowPluginState = function(identity, payload) {
+    const { uuid, name } = identity;
+    const { name: pluginName, version, state, allDone } = payload;
+    const updateTopic = allDone ? 'plugin-state-changed' : 'plugin-state-changing';
+    let { pluginState } = Window.wrap(uuid, name);
+
+    // Single plugin state change
+    if (!allDone) {
+        pluginState = pluginState.find(e => e.name === pluginName && e.version === version);
```
Comment (datamadic) [inline src/browser/api/window.js]: what is `e` short for, lets maybe just call it that? (`pluginState` does not sound like an array, not sure what would be in it...)

File: src/browser/api/window.js [hunk 3]
```diff
@@ -1137,26 +1140,39 @@ Window.getParentApplication = function() {
...
-    let preloadState = openfinWindow.preloadState;
-    const preloadStateUpdateTopic = allDone ? 'preload-state-changed' : 'preload-state-changing';
+    const { url, state, allDone } = payload;
+    const updateTopic = allDone ? 'preload-state-changed' : 'preload-state-changing';
+    let { preloadState } = Window.wrap(uuid, name);
 
     // Single preload script state change
     if (!allDone) {
-        preloadState = preloadState.find(e => e.url === getIdentifier(payload));
+        preloadState = preloadState.find(e => e.url === url);
```
Comment (datamadic) [inline src/browser/api/window.js]: again, the `e`

File: src/browser/api_protocol/api_handlers/window.js [hunk 4]
```diff
@@ -133,6 +134,14 @@ function setWindowBounds(identity, message, ack) {
     ack(successAck);
 }
 
+function setWindowPluginState(identity, message, ack) {
+    const payload = message.payload;
```
Comment (datamadic) [inline src/browser/api_protocol/api_handlers/window.js]: destructure?

File: src/renderer/api-decorator.js [hunk 5]
```diff
@@ -560,52 +560,93 @@ limitations under the License.
...
+        const windowOptions = getWindowOptionsSync();
 
-        if (preloadOption.length) { // short-circuit
-            let response;
-            try {
-                response = syncApiCall('get-selected-preload-scripts', preloadOption);
-            } catch (error) {
-                logPreload('error', identity, 'error', '', error);
-            }
+        let { plugin, preload } = convertOptionsToElectronSync(windowOptions);
```
Comment (datamadic) [inline src/renderer/api-decorator.js]: is `convertOptionsToElectronSync` safe to call multiple times? Is it called elsewhere?

File: src/renderer/api-decorator.js [hunk 6]
```diff
@@ -560,52 +560,93 @@ limitations under the License.
...
+            return syncApiCall('write-to-log', { level: 'error', message: logBase + error });
         }
-        const message = `[PRELOAD] [${identity.uuid}]-[${identity.name}] ${state}`;
-        syncApiCall('write-to-log', { level, message });
+
+        preloadScripts.forEach((preloadScript) => {
+            const { url, content } = preloadScript;
+
+            if (content !== null) {
+                // TODO: handle empty script for bad urls
```
Comment (datamadic) [inline src/renderer/api-decorator.js]: What happens currently, just that failed eval below? (just asking)

File: src/browser/api/window.js (hunk 2 above)
Comment (nicholasdgoodman) [reply src/browser/api/window.js]: e is for entry?

File: src/browser/api_protocol/api_handlers/window.js (hunk 4 above)
Comment (whyn07m3) [reply src/browser/api_protocol/api_handlers/window.js]: this one is my favorite :)

File: src/renderer/api-decorator.js (hunk 5 above)
Comment (whyn07m3) [reply src/renderer/api-decorator.js]: it was there before the change... might change later on, not in this PR (phase 1)

File: src/renderer/api-decorator.js (hunk 6 above)
Comment (whyn07m3) [reply src/renderer/api-decorator.js]: right now this is not working (with and without change)... like many other things... will be fixing

File: src/browser/api_protocol/api_handlers/window.js (hunk 4 above)
Comment (datamadic) [reply src/browser/api_protocol/api_handlers/window.js]: the student has become the teacher

**Comment by whyn07m3** [comment]
✅ Test results after rebasing:
• [Windows 7](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59ef66b7af26a25be2ffc927)
• [Windows 10](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59ef671aaf26a25be2ffc928)

**Pull request merged**